File: WMS.DALCore/Trans_re/clsLnTrans_re_det_lote_num.cs
using System.Data;
using System.Reflection;
using Microsoft.Data.SqlClient;
using Microsoft.Extensions.Configuration;
using Microsoft.VisualBasic.CompilerServices;
using WMSWebAPI.Be;

public class clsLnTrans_re_det_lote_num
{

    private static clsInsert Ins = new clsInsert();
    private static clsUpdate Upd = new clsUpdate();

    public static void Cargar(ref clsBeTrans_re_det_lote_num oBeTrans_re_det_lote_num, DataRow dr)
    {
        int GetInt(string col) { return dr[col] is DBNull ? 0 : Convert.ToInt32(dr[col]); }        
        string GetString(string col) { return dr[col] is DBNull ? "" : (Convert.ToString(dr[col]) ?? ""); }
        DateTime GetDate(string col) { return dr[col] is DBNull ? DateTime.Now : Convert.ToDateTime(dr[col]); }        
        double GetDecimal(string col) { return dr[col] is DBNull ? 0 : Convert.ToDouble(dr[col]); }

        try
        {
            oBeTrans_re_det_lote_num.IdLoteNum = GetInt("IdLoteNum");
            oBeTrans_re_det_lote_num.IdRecepcionEnc = GetInt("IdRecepcionEnc");
            oBeTrans_re_det_lote_num.IdProductoBodega = GetInt("IdProductoBodega");
            oBeTrans_re_det_lote_num.Codigo = GetString("Codigo");
            oBeTrans_re_det_lote_num.Lote = GetString("Lote");
            oBeTrans_re_det_lote_num.Lote_Numerico = GetInt("Lote_Numerico");
            oBeTrans_re_det_lote_num.FechaIngreso = GetDate("FechaIngreso");
            oBeTrans_re_det_lote_num.Cantidad = GetDecimal("Cantidad");
        }
        catch (Exception)
        {
            throw;
        }
    }

    public static int Insertar(clsBeTrans_re_det_lote_num oBeTrans_re_det_lote_num, SqlConnection pConection, SqlTransaction pTransaction)
    {
        SqlCommand cmd = new SqlCommand();

        try
        {
            Ins.Init("trans_re_det_lote_num");
            Ins.Add("idlotenum", "@idlotenum", "F");
            Ins.Add("idrecepcionenc", "@idrecepcionenc", "F");
            Ins.Add("idproductobodega", "@idproductobodega", "F");
            Ins.Add("codigo", "@codigo", "F");
            Ins.Add("lote", "@lote", "F");
            Ins.Add("lote_numerico", "@lote_numerico", "F");
            Ins.Add("fechaingreso", "@fechaingreso", "F");
            Ins.Add("cantidad", "@cantidad", "F");

            string sp = Ins.SQL();
            cmd = new SqlCommand(sp, pConection, pTransaction) { CommandType = CommandType.Text };

            cmd.Parameters.Add(new SqlParameter("@IdLoteNum", oBeTrans_re_det_lote_num.IdLoteNum));
            cmd.Parameters.Add(new SqlParameter("@IdRecepcionEnc", oBeTrans_re_det_lote_num.IdRecepcionEnc));
            cmd.Parameters.Add(new SqlParameter("@IdProductoBodega", oBeTrans_re_det_lote_num.IdProductoBodega));
            cmd.Parameters.Add(new SqlParameter("@Codigo", oBeTrans_re_det_lote_num.Codigo));
            cmd.Parameters.Add(new SqlParameter("@Lote", oBeTrans_re_det_lote_num.Lote));
            cmd.Parameters.Add(new SqlParameter("@Lote_Numerico", oBeTrans_re_det_lote_num.Lote_Numerico));
            cmd.Parameters.Add(new SqlParameter("@FechaIngreso", oBeTrans_re_det_lote_num.FechaIngreso));
            cmd.Parameters.Add(new SqlParameter("@Cantidad", oBeTrans_re_det_lote_num.Cantidad));

            int rowsAffected = cmd.ExecuteNonQuery();
            return rowsAffected;
        }
        catch (Exception)
        {
            throw;
        }
        finally
        {
            cmd?.Dispose();
        }
    }

    public static int Insertar(IConfiguration config, clsBeTrans_re_det_lote_num oBeTrans_re_det_lote_num)
    {

        int rowsAffected = 0;
        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            Ins.Init("trans_re_det_lote_num");
            Ins.Add("idlotenum", "@idlotenum", "F");
            Ins.Add("idrecepcionenc", "@idrecepcionenc", "F");
            Ins.Add("idproductobodega", "@idproductobodega", "F");
            Ins.Add("codigo", "@codigo", "F");
            Ins.Add("lote", "@lote", "F");
            Ins.Add("lote_numerico", "@lote_numerico", "F");
            Ins.Add("fechaingreso", "@fechaingreso", "F");
            Ins.Add("cantidad", "@cantidad", "F");

            string sp = Ins.SQL();

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
            cmd = new SqlCommand(sp, lConnection, lTransaction);

            cmd.Parameters.Add(new SqlParameter("@IdLoteNum", oBeTrans_re_det_lote_num.IdLoteNum));
            cmd.Parameters.Add(new SqlParameter("@IdRecepcionEnc", oBeTrans_re_det_lote_num.IdRecepcionEnc));
            cmd.Parameters.Add(new SqlParameter("@IdProductoBodega", oBeTrans_re_det_lote_num.IdProductoBodega));
            cmd.Parameters.Add(new SqlParameter("@Codigo", oBeTrans_re_det_lote_num.Codigo));
            cmd.Parameters.Add(new SqlParameter("@Lote", oBeTrans_re_det_lote_num.Lote));
            cmd.Parameters.Add(new SqlParameter("@Lote_Numerico", oBeTrans_re_det_lote_num.Lote_Numerico));
            cmd.Parameters.Add(new SqlParameter("@FechaIngreso", oBeTrans_re_det_lote_num.FechaIngreso));
            cmd.Parameters.Add(new SqlParameter("@Cantidad", oBeTrans_re_det_lote_num.Cantidad));

            rowsAffected = cmd.ExecuteNonQuery();

            if (lTransaction != null)
                lTransaction.Commit();

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return rowsAffected;
    }

    public static int Actualizar(clsBeTrans_re_det_lote_num oBeTrans_re_det_lote_num, SqlConnection pConection, SqlTransaction pTransaction)
    {
        SqlCommand cmd = new SqlCommand();

        try
        {
            Upd.Init("trans_re_det_lote_num");
            Upd.Add("idlotenum", "@idlotenum", "F");
            Upd.Add("idrecepcionenc", "@idrecepcionenc", "F");
            Upd.Add("idproductobodega", "@idproductobodega", "F");
            Upd.Add("codigo", "@codigo", "F");
            Upd.Add("lote", "@lote", "F");
            Upd.Add("lote_numerico", "@lote_numerico", "F");
            Upd.Add("fechaingreso", "@fechaingreso", "F");
            Upd.Add("cantidad", "@cantidad", "F");
            Upd.Where("IdLoteNum = @IdLoteNum");

            string sp = Upd.SQL();
            cmd = new SqlCommand(sp, pConection, pTransaction) { CommandType = CommandType.Text };

            cmd.Parameters.Add(new SqlParameter("@IdLoteNum", oBeTrans_re_det_lote_num.IdLoteNum));
            cmd.Parameters.Add(new SqlParameter("@IdRecepcionEnc", oBeTrans_re_det_lote_num.IdRecepcionEnc));
            cmd.Parameters.Add(new SqlParameter("@IdProductoBodega", oBeTrans_re_det_lote_num.IdProductoBodega));
            cmd.Parameters.Add(new SqlParameter("@Codigo", oBeTrans_re_det_lote_num.Codigo));
            cmd.Parameters.Add(new SqlParameter("@Lote", oBeTrans_re_det_lote_num.Lote));
            cmd.Parameters.Add(new SqlParameter("@Lote_Numerico", oBeTrans_re_det_lote_num.Lote_Numerico));
            cmd.Parameters.Add(new SqlParameter("@FechaIngreso", oBeTrans_re_det_lote_num.FechaIngreso));
            cmd.Parameters.Add(new SqlParameter("@Cantidad", oBeTrans_re_det_lote_num.Cantidad));

            int rowsAffected = cmd.ExecuteNonQuery();
            return rowsAffected;
        }
        catch (Exception)
        {
            throw;
        }
        finally
        {
            cmd?.Dispose();
        }
    }

    public int Eliminar(IConfiguration config, clsBeTrans_re_det_lote_num oBeTrans_re_det_lote_num, SqlConnection? pConection = null, SqlTransaction? pTransaction = null)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {
            const string sp = (" Delete from Trans_re_det_lote_num" +
             "  Where(IdLoteNum = @IdLoteNum)");

            bool Es_Transaccion_Remota = (pConection != null && pTransaction != null);

            SqlCommand cmd = new SqlCommand() { CommandType = CommandType.Text };

            if (Es_Transaccion_Remota)
            {
                cmd = new SqlCommand(sp, pConection, pTransaction);
            }
            else
            {
                lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);
                cmd = new SqlCommand(sp, lConnection, lTransaction);
            }

            cmd.Parameters.Add(new SqlParameter("@IdLoteNum", oBeTrans_re_det_lote_num.IdLoteNum));

            int rowsAffected = cmd.ExecuteNonQuery();

            if (!Es_Transaccion_Remota)
                if (lTransaction != null)
                    lTransaction.Commit();

            return rowsAffected;

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
    }    

    public static bool GetSingle(IConfiguration config, ref clsBeTrans_re_det_lote_num pBeTrans_re_det_lote_num)
    {

        SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTransaction = null;

        try
        {

            const string sp = "Select * FROM Trans_re_det_lote_num" +
            " Where(IdLoteNum = @IdLoteNum)";

            lConnection.Open(); lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted);

            SqlCommand cmd = new SqlCommand(sp, lConnection, lTransaction) { CommandType = CommandType.Text };
            SqlDataAdapter dad = new SqlDataAdapter(cmd);

            dad.SelectCommand.Parameters.Add(new SqlParameter("@IdLoteNum", pBeTrans_re_det_lote_num.IdLoteNum));            

            DataTable dt = new DataTable();
            dad.Fill(dt);

            lTransaction.Commit();

            if (dt.Rows.Count == 1)
            {
                DataRow r;
                r = dt.Rows[0];
                Cargar(ref pBeTrans_re_det_lote_num, r);
                return true;
            }

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
        finally
        {
            if (lConnection.State == ConnectionState.Open) lConnection.Close();
            if (lConnection != null) lConnection.Dispose();
            if (lTransaction != null) lTransaction.Dispose();
        }
        return false;

    }

    public static List<clsBeTrans_re_det_lote_num> GetAll(IConfiguration config)
    {

        SqlTransaction? lTransaction = null;
        List<clsBeTrans_re_det_lote_num> lreturnList = new List<clsBeTrans_re_det_lote_num>();

        try
        {
            const string sp = "Select * FROM Trans_re_det_lote_num";

            using (SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST")))
            {

                lConnection.Open();

                using (lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted))
                {
                    using (SqlDataAdapter lDTA = new SqlDataAdapter(sp, lConnection))
                    {
                        lDTA.SelectCommand.CommandType = CommandType.Text;
                        lDTA.SelectCommand.Transaction = lTransaction;
                        DataTable lDataTable = new DataTable();
                        lDTA.Fill(lDataTable);

                        clsBeTrans_re_det_lote_num vBeTrans_re_det_lote_num = new clsBeTrans_re_det_lote_num();

                        foreach (DataRow dr in lDataTable.Rows)
                        {
                            vBeTrans_re_det_lote_num = new clsBeTrans_re_det_lote_num();
                            Cargar(ref vBeTrans_re_det_lote_num, dr);
                            lreturnList.Add(vBeTrans_re_det_lote_num);
                        }

                        lTransaction.Commit();
                    }

                    lConnection.Close();

                }

            }

            return lreturnList;

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
    }

    public static int MaxID(IConfiguration config)
    {

        SqlTransaction? lTransaction = null;

        try
        {

            int lMax = 0;

            const string sp = "Select ISNULL(Max(IdLoteNum),0) FROM Trans_re_det_lote_num";

            using (SqlConnection lConnection = new SqlConnection(config.GetConnectionString("CST")))
            {
                lConnection.Open();

                using (lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadUncommitted))
                {
                    using (SqlCommand lCommand = new SqlCommand(sp, lConnection) { CommandType = CommandType.Text })
                    {
                        var lreturnValue = lCommand.ExecuteScalar();
                        if (lreturnValue != DBNull.Value && lreturnValue != null)
                        {
                            lMax = int.Parse((string)lreturnValue);
                        }
                    }
                    lTransaction.Commit();
                }

                lConnection.Close();
            }

            return lMax;

        }
        catch (SqlException)
        {
            if (lTransaction is not null)
                lTransaction.Rollback();
            throw;
        }
    }
    public static int MaxID(SqlConnection pConection, SqlTransaction pTransaction)
    {
        try
        {
            const string sp = "Select ISNULL(Max(IdLoteNum),0) FROM Trans_re_det_lote_num";

            using (var cmd = new SqlCommand(sp, pConection, pTransaction) { CommandType = CommandType.Text })
            {
                object lreturnValue = cmd.ExecuteScalar();

                if (lreturnValue != DBNull.Value && lreturnValue != null)
                {
                    return Convert.ToInt32(lreturnValue);
                }
            }

            return 0;
        }
        catch (Exception)
        {
            throw;
        }
    }

}

File: TOMIMSV4/DAL/Transacciones/Recepcion_Detalle_Lote_Num/clsLnTrans_re_det_lote_num.vb
Imports System.Data.SqlClient
Imports System.Reflection

Public Class clsLnTrans_re_det_lote_num

    Public Shared Sub Cargar(ByRef oBeTrans_re_det_lote_num As clsBeTrans_re_det_lote_num, ByRef dr As DataRow)
        Try
            With oBeTrans_re_det_lote_num
                .IdLoteNum = IIf(IsDBNull(dr.Item("IdLoteNum")), 0, dr.Item("IdLoteNum"))
                .IdRecepcionEnc = IIf(IsDBNull(dr.Item("IdRecepcionEnc")), 0, dr.Item("IdRecepcionEnc"))
                .IdProductoBodega = IIf(IsDBNull(dr.Item("IdProductoBodega")), 0, dr.Item("IdProductoBodega"))
                .Codigo = IIf(IsDBNull(dr.Item("Codigo")), "", dr.Item("Codigo"))
                .Lote = IIf(IsDBNull(dr.Item("Lote")), "", dr.Item("Lote"))
                .Lote_Numerico = IIf(IsDBNull(dr.Item("lote_numerico")), 0, dr.Item("lote_numerico"))
                .FechaIngreso = IIf(IsDBNull(dr.Item("FechaIngreso")), Nothing, dr.Item("FechaIngreso"))
                .Cantidad = IIf(IsDBNull(dr.Item("Cantidad")), 0.0, dr.Item("Cantidad"))
            End With
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try
    End Sub
    Public Shared Function Insertar(ByRef oBeTrans_re_det_lote_num As clsBeTrans_re_det_lote_num, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Ins.Init("trans_re_det_lote_num")
            Ins.Add("idlotenum", "@idlotenum", DataType.Parametro)
            Ins.Add("idrecepcionenc", "@idrecepcionenc", DataType.Parametro)
            Ins.Add("idproductobodega", "@idproductobodega", DataType.Parametro)
            Ins.Add("codigo", "@codigo", DataType.Parametro)
            Ins.Add("lote", "@lote", DataType.Parametro)
            Ins.Add("lote_numerico", "@lote_numerico", DataType.Parametro)
            Ins.Add("fechaingreso", "@fechaingreso", DataType.Parametro)
            Ins.Add("cantidad", "@cantidad", DataType.Parametro)

            Dim sp As String = Ins.SQL()
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDLOTENUM", oBeTrans_re_det_lote_num.IdLoteNum))
            cmd.Parameters.Add(New SqlParameter("@IDRECEPCIONENC", oBeTrans_re_det_lote_num.IdRecepcionEnc))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOBODEGA", oBeTrans_re_det_lote_num.IdProductoBodega))
            cmd.Parameters.Add(New SqlParameter("@CODIGO", oBeTrans_re_det_lote_num.Codigo))
            cmd.Parameters.Add(New SqlParameter("@LOTE", oBeTrans_re_det_lote_num.Lote))
            cmd.Parameters.Add(New SqlParameter("@LOTE_NUMERICO", oBeTrans_re_det_lote_num.Lote_Numerico))
            cmd.Parameters.Add(New SqlParameter("@FECHAINGRESO", oBeTrans_re_det_lote_num.FechaIngreso))
            cmd.Parameters.Add(New SqlParameter("@CANTIDAD", oBeTrans_re_det_lote_num.Cantidad))


            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
        End Try

    End Function
    Public Shared Function Actualizar(ByRef oBeTrans_re_det_lote_num As clsBeTrans_re_det_lote_num, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Upd.Init("trans_re_det_lote_num")
            Upd.Add("idlotenum", "@idlotenum", DataType.Parametro)
            Upd.Add("idrecepcionenc", "@idrecepcionenc", DataType.Parametro)
            Upd.Add("idproductobodega", "@idproductobodega", DataType.Parametro)
            Upd.Add("codigo", "@codigo", DataType.Parametro)
            Upd.Add("lote", "@lote", DataType.Parametro)
            Upd.Add("lote_numerico", "@lote_numerico", DataType.Parametro)
            Upd.Add("fechaingreso", "@fechaingreso", DataType.Parametro)
            Upd.Add("cantidad", "@cantidad", DataType.Parametro)
            Upd.Where("IdLoteNum = @IdLoteNum")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}


            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDLOTENUM", oBeTrans_re_det_lote_num.IdLoteNum))
            cmd.Parameters.Add(New SqlParameter("@IDRECEPCIONENC", oBeTrans_re_det_lote_num.IdRecepcionEnc))
            cmd.Parameters.Add(New SqlParameter("@IDPRODUCTOBODEGA", oBeTrans_re_det_lote_num.IdProductoBodega))
            cmd.Parameters.Add(New SqlParameter("@CODIGO", oBeTrans_re_det_lote_num.Codigo))
            cmd.Parameters.Add(New SqlParameter("@LOTE", oBeTrans_re_det_lote_num.Lote))
            cmd.Parameters.Add(New SqlParameter("@LOTE_NUMERICO", oBeTrans_re_det_lote_num.Lote_Numerico))
            cmd.Parameters.Add(New SqlParameter("@FECHAINGRESO", oBeTrans_re_det_lote_num.FechaIngreso))
            cmd.Parameters.Add(New SqlParameter("@CANTIDAD", oBeTrans_re_det_lote_num.Cantidad))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
        End Try

    End Function
    Public Shared Function Eliminar(ByRef oBeTrans_re_det_lote_num As clsBeTrans_re_det_lote_num, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Trans_re_det_lote_num" &
             "  Where(IdLoteNum = @IdLoteNum)"

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand With {.CommandType = CommandType.Text}

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
                cmd.Transaction = pTransaction
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDLOTENUM", oBeTrans_re_det_lote_num.IdLoteNum))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
        End Try

    End Function
    Public Shared Function EliminarTodos(Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Trans_re_det_lote_num"
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            If lTransaction IsNot Nothing Then lTransaction.Dispose()
        End Try

    End Function
    Public Shared Function Listar() As DataTable

        Try

            Const sp As String = "SELECT * FROM Trans_re_det_lote_num"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))

            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable
            dad.Fill(dt)

            Return dt

        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function
    Public Shared Function Obtener(ByRef oBeTrans_re_det_lote_num As clsBeTrans_re_det_lote_num) As Boolean

        Try

            Const sp As String = "SELECT * FROM Trans_re_det_lote_num" &
            " Where(IdLoteNum = @IdLoteNum)"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))

            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)

            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDLOTENUM", oBeTrans_re_det_lote_num.IdLoteNum))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(oBeTrans_re_det_lote_num, dt.Rows(0))
            End If

            Return True

        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function
    Public Shared Function GetAll() As List(Of clsBeTrans_re_det_lote_num)

        Try

            Dim lReturnList As New List(Of clsBeTrans_re_det_lote_num)
            Const sp As String = "SELECT * FROM Trans_re_det_lote_num"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))

            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable

            dad.Fill(dt)

            Dim vBeTrans_re_det_lote_num As New clsBeTrans_re_det_lote_num

            For Each dr As DataRow In dt.Rows
                vBeTrans_re_det_lote_num = New clsBeTrans_re_det_lote_num
                Cargar(vBeTrans_re_det_lote_num, dr)
                lReturnList.Add(vBeTrans_re_det_lote_num)
            Next

            Return lReturnList

            If lConnection.State = ConnectionState.Open Then lConnection.Close()
            lConnection.Dispose()
            cmd.Dispose()

        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function
    Public Shared Function GetSingle(ByRef pBeTrans_re_det_lote_num As clsBeTrans_re_det_lote_num)

        Try

            Const sp As String = "SELECT * FROM Trans_re_det_lote_num" &
            " Where(IdLoteNum = @IdLoteNum)"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))

            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDLOTENUM", pBeTrans_re_det_lote_num.IDLOTENUM))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(pBeTrans_re_det_lote_num, dt.Rows(0))
            End If

            Return True

        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try

    End Function
    Public Shared Function MaxID() as Integer

        Try

            Dim lMax As Integer = 0

            Const sp As String = "SELECT ISNULL(Max(IdLoteNum),0) FROM Trans_re_det_lote_num"

            Using lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
                lConnection.Open()
                Using lCommand As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
                    Dim lReturnValue As Object = lCommand.ExecuteScalar()
                    lConnection.Close()
                    If lReturnValue IsNot DBNull.Value AndAlso lReturnValue IsNot Nothing Then
                        lMax = CInt(lReturnValue)
                    End If
                End Using
            End Using

            Return lMax

        Catch ex As Exception
            Throw New Exception(String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message))
        End Try

    End Function

End Class